packages/cli/src/utils/tunnel-utils.ts

```diff
@@ -0,0 +1,141 @@
+import { TunnelTail, ClientCloseEvent, ClientErrorEvent } from '@bpinternal/tunnel'
+import { Logger } from '../logger'
+import { EventEmitter } from './event-emitter'
+
+export type ReconnectionTriggerEvent =
+  | {
+      type: 'init'
+      ev: null
+    }
+  | {
+      type: 'error'
+      ev: ClientErrorEvent
+    }
+  | {
+      type: 'close'
+      ev: ClientCloseEvent
+    }
+
+export type ReconnectedEvent = {
+  tunnel: TunnelTail
+  ev: ReconnectionTriggerEvent
+}
+
+export class ReconnectionFailedError extends Error {
+  public constructor(public readonly event: ReconnectionTriggerEvent) {
+    const reason = ReconnectionFailedError._reason(event)
+    super(`Reconnection failed: ${reason}`)
+  }
+
+  private static _reason(event: ReconnectionTriggerEvent): string {
+    if (event.type === 'error') {
+      return 'error'
+    }
+
+    if (event.type === 'close') {
+      return `${event.ev.code} ${event.ev.reason}`
+    }
+
+    return 'init'
+  }
+}
+
+export class TunnelSupervisor {
+  private _tunnel?: TunnelTail
+  private _closed = false
+  public readonly events = new EventEmitter<{
+    connectionFailed: ReconnectionTriggerEvent
+    manuallyClosed: null
+    connected: {
+      tunnel: TunnelTail
+      ev: ReconnectionTriggerEvent
+    }
+  }>()
+
+  constructor(private _tunnelUrl: string, private _tunnelId: string, private _logger: Logger) {}
+
+  public async start(): Promise<void> {
+    if (this._closed) {
+      throw new Error('Tunnel is closed')
+    }
+    if (this._tunnel) {
+      throw new Error('Tunnel is already started')
+    }
+
+    const tunnel = await this._reconnect({ type: 'init', ev: null })
+    this._tunnel = tunnel
```

Comment:
**🎯 suggestion:** ‏ We might want to use a separate boolean flag (`tunnelStarted`) that we can set to `true` as a faster locking mechanism before calling `._reconnect()`, because if `start()` is called more than once near the same time for some reason (e.g. a bug) then `_reconnect()` might not resolve fast enough and we'll end up with multiple connections?

Reply:
Yes you're right; there is a race condition. Altough its not to bad because this class is not public, I still want to make sure that its consistent. I'll fix this